### PR TITLE
NuGet feed deployment repository

### PR DESCRIPTION
**Pull request by jkonecki**
Ability to publish application binaries as NuGet feed.

 - [x] Extract `IDeploymentConfigRepository` from `IDeploymentRepository`
 - [x] Implement `NugetFeedDeploymentRepository` for downloading application binaries (separate project)
 - [x] Implement `NugetPackageExtractor` (could be used to extract NuGet packages uploaded to blob storage)

### CODE REVIEW

File: src/NugetFeedDeploymentRepository/Logger.cs [hunk 1]
```diff
@@ -0,0 +1,94 @@
+﻿using NuGet.Common;
+using System;
+using System.Diagnostics;
+using System.Threading.Tasks;
+
+namespace Etg.Yams.NuGet.Storage
+{
+    public class Logger : ILogger
```
Comment (nehmebilal) [inline src/NugetFeedDeploymentRepository/Logger.cs]: This is an implementation of the adapter pattern to adapt Trace to the ILogger interface. It would be good if the name can reflect that (e.g. TraceToNuGetILoggerAdapter).

File: src/NugetFeedDeploymentRepository/NugetFeedDeploymentRepository.cs [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+            PackageMetadataResource packageMetadataResource = await _sourceRepository.GetResourceAsync<PackageMetadataResource>();
+            PackageIdentity packageIdentity = new PackageIdentity(appIdentity.Id, new NuGetVersion(appIdentity.Version.ToString()));
+            IPackageSearchMetadata searchMetadata = await packageMetadataResource.GetMetadataAsync(packageIdentity, new SourceCacheContext(), new Logger(), CancellationToken.None);
+
+            return searchMetadata?.IsListed ?? false;
+        }
+
+        public Task UploadApplicationBinaries(AppIdentity appIdentity, string localPath, ConflictResolutionMode conflictResolutionMode)
+        {
+            throw new NotSupportedException("Yams doesn't support publishing NuGet packages. Publish your packages from your build server.");
```
Comment (nehmebilal) [inline src/NugetFeedDeploymentRepository/NugetFeedDeploymentRepository.cs]: > Yams doesn't support publishing NuGet packages

It's more that this repository doesn't support publishing. In fact, why not? Users of `IDeploymentRepository` that are using it to upload binaries will not be able to use this plugin.

File: src/Etg.Yams.Core/Storage/IDeploymentRepository.cs [hunk 3]
```diff
@@ -1,13 +1,10 @@
 ﻿using System.Threading.Tasks;
 using Etg.Yams.Application;
-using Etg.Yams.Storage.Config;
 
 namespace Etg.Yams.Storage
 {
     public interface IDeploymentRepository
```
Comment (nehmebilal) [inline src/Etg.Yams.Core/Storage/IDeploymentRepository.cs]: What about:
```csharp
public interface IDeploymentRepository : IDeploymentConfigRepository, IApplicationsRepository
```

This would allow you to keep the current Yams Api as is and avoid breaking any users (not to mention a lot of documentation to update). You can add a class to compose an IDeploymentRepository implementation from IDeploymentConfigRepository and IApplicationsRepository:

```csharp
public class DeploymentRepository
{
   public DeploymentRepository(IDeploymentConfigRepository deploymentConfigRepository, IApplicationsRepository applicationsRepository)
   {
   }
}
```

File: test/NugetFeedDeploymentRepositoryTest/NugetFeedDeploymentRepositoryTest.cs [hunk 4]
```diff
@@ -0,0 +1,33 @@
+﻿using System;
+using System.Threading.Tasks;
+using Etg.Yams.Application;
+using Etg.Yams.NuGet.Storage;
+using Xunit;
+
+namespace Etg.Yams.Nuget.Storage.Test
+{
+    [Trait("Category", "Integration")]
+    public class NugetFeedDeploymentRepositoryTest
```
[the remark this answers is not in the document]
Comment (jkonecki) [reply test/NugetFeedDeploymentRepositoryTest/NugetFeedDeploymentRepositoryTest.cs]: Not finished yet ;-)
image

File: src/NugetFeedDeploymentRepository/Logger.cs (hunk 1 above)
Comment (jkonecki) [reply src/NugetFeedDeploymentRepository/Logger.cs]: Renaming to `TraceLogger`

File: src/NugetFeedDeploymentRepository/NugetFeedDeploymentRepository.cs (hunk 2 above)
Comment (jkonecki) [reply src/NugetFeedDeploymentRepository/NugetFeedDeploymentRepository.cs]: I don't think the upload feature is needed that much and it will be more complex due to the need for credentials. I was hoping I could implement it later.

File: src/Etg.Yams.Core/Storage/IDeploymentRepository.cs (hunk 3 above)
Comment (jkonecki) [reply src/Etg.Yams.Core/Storage/IDeploymentRepository.cs]: I was thinking about introducing `IApplicationRepository` or even separate interfaces for uploading and downloading binaries. Will implement the above and await your opinion on the need of pacjage publishing through Yams.

File: src/NugetFeedDeploymentRepository/NugetFeedDeploymentRepository.cs (hunk 2 above)
Comment (nehmebilal) [reply src/NugetFeedDeploymentRepository/NugetFeedDeploymentRepository.cs]: That's fine. Just to give an example, the `InstallApplicationsCmdlet` calls the upload Api and will not work with this repository if it doesn't support publishing.

**Comment by jkonecki** [comment]
I will take a look how complex this is. Maybe it won't be that bad...

File: src/NugetFeedDeploymentRepository/NugetFeedApplicationRepository.cs [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+using NuGet.Protocol.Core.Types;
+using NuGet.Versioning;
+
+namespace Etg.Yams.NuGet.Storage
+{
+    public class NugetFeedApplicationRepository : IApplicationRepository
+    {
+        private readonly SourceRepository _sourceRepository;
+
+        public NugetFeedApplicationRepository(string feedUrl = "https://api.nuget.org/v3/index.json", NugetFeedCredentials credentials = null)
```
Comment (nehmebilal) [inline src/NugetFeedDeploymentRepository/NugetFeedApplicationRepository.cs]: please extract default value to a const

File: src/NugetFeedDeploymentRepository/NugetFeedApplicationRepository.cs [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+        {
+            this.FeedUrl = feedUrl;
+
+            List<Lazy<INuGetResourceProvider>> providers = new List<Lazy<INuGetResourceProvider>>();
+            providers.AddRange(Repository.Provider.GetCoreV3());
+            PackageSource packageSource = new PackageSource(feedUrl);
+
+            if(credentials != null)
```
Comment (nehmebilal) [inline src/NugetFeedDeploymentRepository/NugetFeedApplicationRepository.cs]: Please add brackets to if statement body

File: src/NugetFeedDeploymentRepository/NugetFeedApplicationRepository.cs [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+        }
+
+        public string FeedUrl { get; private set; }
+
+        public Task DeleteApplicationBinaries(AppIdentity appIdentity)
+        {
+            throw new NotSupportedException("NuGet feeds are immutable.");
+        }
+
+        public async Task DownloadApplicationBinaries(AppIdentity appIdentity, string localPath, ConflictResolutionMode conflictResolutionMode)
```
Comment (nehmebilal) [inline src/NugetFeedDeploymentRepository/NugetFeedApplicationRepository.cs]: It doesn't look like the conflictResolutionMode is being handled?

File: src/NugetFeedDeploymentRepository/NugetFeedApplicationRepository.cs [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+        {
+            string tempPath = Path.GetTempPath();
+            DownloadResource downloadResource = await _sourceRepository.GetResourceAsync<DownloadResource>();
+            PackageIdentity packageIdentity = new PackageIdentity(appIdentity.Id, new NuGetVersion(appIdentity.Version.ToString()));
+            PackageDownloadContext downloadContext = new PackageDownloadContext(new SourceCacheContext(), tempPath, true);
+
+            var result = await downloadResource.GetDownloadResourceResultAsync(packageIdentity, downloadContext, tempPath, new TraceLogger(), CancellationToken.None);
+
+            if (result.Status != DownloadResourceResultStatus.Available)
```
Comment (nehmebilal) [inline src/NugetFeedDeploymentRepository/NugetFeedApplicationRepository.cs]: Please add brackets to if

File: src/NugetFeedDeploymentRepository/NugetFeedApplicationRepository.cs [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+        }
+
+        public async Task DownloadApplicationBinaries(AppIdentity appIdentity, string localPath, ConflictResolutionMode conflictResolutionMode)
+        {
+            string tempPath = Path.GetTempPath();
+            DownloadResource downloadResource = await _sourceRepository.GetResourceAsync<DownloadResource>();
+            PackageIdentity packageIdentity = new PackageIdentity(appIdentity.Id, new NuGetVersion(appIdentity.Version.ToString()));
+            PackageDownloadContext downloadContext = new PackageDownloadContext(new SourceCacheContext(), tempPath, true);
+
+            var result = await downloadResource.GetDownloadResourceResultAsync(packageIdentity, downloadContext, tempPath, new TraceLogger(), CancellationToken.None);
```
Comment (nehmebilal) [inline src/NugetFeedDeploymentRepository/NugetFeedApplicationRepository.cs]: Make TraceLogger a field?

Comment (nehmebilal) [inline src/NugetFeedDeploymentRepository/NugetFeedApplicationRepository.cs]: Please only use var when return type is obvious

File: src/NugetFeedDeploymentRepository/NugetFeedApplicationRepository.cs [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+        {
+            throw new NotSupportedException("NuGet feeds are immutable.");
+        }
+
+        public async Task DownloadApplicationBinaries(AppIdentity appIdentity, string localPath, ConflictResolutionMode conflictResolutionMode)
+        {
+            string tempPath = Path.GetTempPath();
+            DownloadResource downloadResource = await _sourceRepository.GetResourceAsync<DownloadResource>();
+            PackageIdentity packageIdentity = new PackageIdentity(appIdentity.Id, new NuGetVersion(appIdentity.Version.ToString()));
+            PackageDownloadContext downloadContext = new PackageDownloadContext(new SourceCacheContext(), tempPath, true);
```
Comment (nehmebilal) [inline src/NugetFeedDeploymentRepository/NugetFeedApplicationRepository.cs]: Please use a named parameter for true to clarify what the parameter is

File: src/NugetFeedDeploymentRepository/NugetFeedCredentials.cs [hunk 11]
```diff
@@ -0,0 +1,14 @@
...
+{
+    public class NugetFeedCredentials
+    {
+        public NugetFeedCredentials(string username, string password)
+        {
+            this.Username = username;
+            this.Password = password;
+        }
+
+        public string Username { get; private set; }
```
Comment (nehmebilal) [inline src/NugetFeedDeploymentRepository/NugetFeedCredentials.cs]: I noticed you always add private set, what for? I would avoid them unless needed.

File: src/NugetFeedDeploymentRepository/NugetPackageExtractor.cs [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using NuGet.Packaging;
+using NuGet.Packaging.Core;
+using NuGet.Packaging.PackageExtraction;
+
+namespace Etg.Yams.NuGet.Storage
+{
+    public class NugetPackageExtractor
```
Comment (nehmebilal) [inline src/NugetFeedDeploymentRepository/NugetPackageExtractor.cs]: Looks like a utils, or even just a private method inside the nuget repository

File: src/NugetFeedDeploymentRepository/NugetPackageExtractor.cs [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+    class FlatPackagePathResolver : PackagePathResolver
+    {
+        public FlatPackagePathResolver(string rootDirectory)
+            : base(rootDirectory, false)
+        {
+        }
+
+        public override string GetPackageDirectoryName(PackageIdentity packageIdentity)
+        {
+            return string.Empty;
```
Comment (nehmebilal) [inline src/NugetFeedDeploymentRepository/NugetPackageExtractor.cs]: a comment would be useful here

File: src/NugetFeedDeploymentRepository/NugetPackageExtractor.cs [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+        }
+    }
+
+    /// <summary>
+    /// Extracts package directly to root directory.
+    /// </summary>
+    class FlatPackagePathResolver : PackagePathResolver
+    {
+        public FlatPackagePathResolver(string rootDirectory)
+            : base(rootDirectory, false)
```
Comment (nehmebilal) [inline src/NugetFeedDeploymentRepository/NugetPackageExtractor.cs]: not sure what is false? A named parameter would help :)

File: src/NugetFeedDeploymentRepository/NugetPackageExtractor.cs [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+        }
+
+        public override string GetInstalledPackageFilePath(PackageIdentity packageIdentity)
+        {
+            return PackagePathHelper.GetInstalledPackageFilePath(packageIdentity, this);
+        }
+
+        private string GetId(PackageIdentity identity)
+        {
+            return string.Empty;
```
Comment (nehmebilal) [inline src/NugetFeedDeploymentRepository/NugetPackageExtractor.cs]: Comment?

File: src/NugetFeedDeploymentRepository/TraceLogger.cs [hunk 16]
```diff
@@ -0,0 +1,94 @@
...
+                    break;
+
+                case LogLevel.Verbose:
+                    this.LogVerbose(data);
+                    break;
+
+                case LogLevel.Information:
+                    this.LogInformation(data);
+                    break;
+            }
```
Comment (nehmebilal) [inline src/NugetFeedDeploymentRepository/TraceLogger.cs]: default: throw

File: test/NugetFeedDeploymentRepositoryTest/NugetFeedApplicationRepositoryTest.cs [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+using System.Threading.Tasks;
+using Etg.Yams.Application;
+using Etg.Yams.NuGet.Storage;
+using Etg.Yams.Storage;
+using Etg.Yams.Utils;
+using Xunit;
+
+namespace Etg.Yams.Nuget.Storage.Test
+{
+    [Trait("Category", "Integration")]
```
Comment (nehmebilal) [inline test/NugetFeedDeploymentRepositoryTest/NugetFeedApplicationRepositoryTest.cs]: This doesn't run in CI. I'd remove the *integration* trait unless there is a reason why the tests should not run as part of the CI build.

File: test/NugetFeedDeploymentRepositoryTest/NugetFeedApplicationRepositoryTest.cs [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+
+        [Fact]
+        public async Task TestDownloadApplicationBinaries_ThrowsWhenPackageNotFound()
+        {
+            AppIdentity appIdentity = new AppIdentity(Guid.NewGuid().ToString(), "1.0.0");
+            string localPath = await CreateTestTempDirectory("Nuget_TestDownloadApplicationBinaries");
+            NugetFeedApplicationRepository _applicationRepository = new NugetFeedApplicationRepository();
+
+            BinariesNotFoundException exception = await Assert.ThrowsAsync<BinariesNotFoundException>(() => _applicationRepository.DownloadApplicationBinaries(appIdentity, localPath, ConflictResolutionMode.OverwriteExistingBinaries));
+            Assert.Contains(appIdentity.ToString(), exception.Message);
```
Comment (nehmebilal) [inline test/NugetFeedDeploymentRepositoryTest/NugetFeedApplicationRepositoryTest.cs]: Testing the exception message feels a bit fragile

File: src/NugetFeedDeploymentRepository/NugetFeedApplicationRepository.cs [hunk 19]
```diff
@@ -0,0 +1,73 @@
...
+    {
+        private readonly SourceRepository _sourceRepository;
+
+        public NugetFeedApplicationRepository(string feedUrl = "https://api.nuget.org/v3/index.json", NugetFeedCredentials credentials = null)
+        {
+            this.FeedUrl = feedUrl;
+
+            List<Lazy<INuGetResourceProvider>> providers = new List<Lazy<INuGetResourceProvider>>();
+            providers.AddRange(Repository.Provider.GetCoreV3());
+            PackageSource packageSource = new PackageSource(feedUrl);
```
Comment (nehmebilal) [inline src/NugetFeedDeploymentRepository/NugetFeedApplicationRepository.cs]: Bonus: Ideally, this would be behind an interface and injected to this class (IoC) so we can use a mock to test edge cases such as what happens if the nuget repo throw a particular exception, the feed is down, etc. It's a bit involved so I'll leave it up to you.

File: src/Etg.Yams.Core/YamsDiModule.cs [hunk 20]
```diff
@@ -20,11 +20,13 @@ public class YamsDiModule
     {
         private readonly IContainer _container;
 
-        public YamsDiModule(YamsConfig config, IDeploymentRepository deploymentRepository,
+        public YamsDiModule(YamsConfig config,
```
Comment (nehmebilal) [inline src/Etg.Yams.Core/YamsDiModule.cs]: I'd keep the old constructor too to stay backward compatible

File: src/Etg.Yams.Core/YamsDiModule.cs [hunk 21]
```diff
@@ -37,8 +39,8 @@ public YamsDiModule(IContainer container)
 
         public IContainer Container => _container;
 
-        public static ContainerBuilder RegisterTypes(YamsConfig config, 
-            IDeploymentRepository deploymentRepository, IDeploymentStatusWriter deploymentStatusWriter,
+        public static ContainerBuilder RegisterTypes(YamsConfig config,
+            IDeploymentConfigRepository deploymentConfigRepository, IApplicationRepository applicationRepository, IDeploymentStatusWriter deploymentStatusWriter,
```
Comment (nehmebilal) [inline src/Etg.Yams.Core/YamsDiModule.cs]: Please keep both old and new versions

File: Etg.Yams.sln [hunk 22]
```diff
@@ -55,6 +55,10 @@ Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Etg.Yams.Ipc", "src\Etg.Yam
 EndProject
 Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Etg.Yams.Host", "src\Etg.Yams.Host\Etg.Yams.Host.csproj", "{C9A993F7-E3E2-4FDC-AA1E-01348B023D16}"
 EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "NugetFeedDeploymentRepository", "src\NugetFeedDeploymentRepository\NugetFeedDeploymentRepository.csproj", "{1110E183-76B6-4C9F-939A-1B9290C3FC72}"
```
Comment (nehmebilal) [inline Etg.Yams.sln]: Are we going to create a NuGet for this repository? I am also wondering if we should create an Yams.Extensions solution (or even potentially a GitHub repo).

File: src/NugetFeedDeploymentRepository/NugetFeedApplicationRepository.cs (hunk 5 above)
Comment (jkonecki) [reply src/NugetFeedDeploymentRepository/NugetFeedApplicationRepository.cs]: Done

File: src/NugetFeedDeploymentRepository/NugetFeedApplicationRepository.cs (hunk 6 above)
Comment (jkonecki) [reply src/NugetFeedDeploymentRepository/NugetFeedApplicationRepository.cs]: Done

File: src/NugetFeedDeploymentRepository/NugetFeedApplicationRepository.cs (hunk 9 above)
Comment (jkonecki) [reply src/NugetFeedDeploymentRepository/NugetFeedApplicationRepository.cs]: Done

Comment (jkonecki) [reply src/NugetFeedDeploymentRepository/NugetFeedApplicationRepository.cs]: Done

File: src/NugetFeedDeploymentRepository/NugetFeedApplicationRepository.cs (hunk 10 above)
Comment (jkonecki) [reply src/NugetFeedDeploymentRepository/NugetFeedApplicationRepository.cs]: Done

File: src/NugetFeedDeploymentRepository/NugetPackageExtractor.cs (hunk 13 above)
Comment (jkonecki) [reply src/NugetFeedDeploymentRepository/NugetPackageExtractor.cs]: Done

File: src/NugetFeedDeploymentRepository/NugetPackageExtractor.cs (hunk 12 above)
Comment (jkonecki) [reply src/NugetFeedDeploymentRepository/NugetPackageExtractor.cs]: I wanted to keep it separate from the repository so it can be easily reused in the future. I was thinking about supporting a scenario where the packages are downloaded from blob storage instead of nuget feed.

File: src/NugetFeedDeploymentRepository/TraceLogger.cs (hunk 16 above)
Comment (jkonecki) [reply src/NugetFeedDeploymentRepository/TraceLogger.cs]: Done

File: src/NugetFeedDeploymentRepository/NugetPackageExtractor.cs (hunk 15 above)
Comment (jkonecki) [reply src/NugetFeedDeploymentRepository/NugetPackageExtractor.cs]: Done

File: src/NugetFeedDeploymentRepository/NugetPackageExtractor.cs (hunk 14 above)
Comment (jkonecki) [reply src/NugetFeedDeploymentRepository/NugetPackageExtractor.cs]: Done

File: test/NugetFeedDeploymentRepositoryTest/NugetFeedApplicationRepositoryTest.cs (hunk 17 above)
Comment (jkonecki) [reply test/NugetFeedDeploymentRepositoryTest/NugetFeedApplicationRepositoryTest.cs]: Done. Didn't know _integration_ means no CI execution.

File: test/NugetFeedDeploymentRepositoryTest/NugetFeedApplicationRepositoryTest.cs (hunk 18 above)
Comment (jkonecki) [reply test/NugetFeedDeploymentRepositoryTest/NugetFeedApplicationRepositoryTest.cs]: I want to ensure that the message contains app identity. I don't validate the whole message, just the important piece of information. I can remove this assertion if you insist.

File: src/Etg.Yams.Core/YamsDiModule.cs (hunk 20 above)
Comment (jkonecki) [reply src/Etg.Yams.Core/YamsDiModule.cs]: Done

File: src/Etg.Yams.Core/YamsDiModule.cs (hunk 21 above)
Comment (jkonecki) [reply src/Etg.Yams.Core/YamsDiModule.cs]: Done

File: src/NugetFeedDeploymentRepository/NugetFeedApplicationRepository.cs (hunk 19 above)
Comment (jkonecki) [reply src/NugetFeedDeploymentRepository/NugetFeedApplicationRepository.cs]: I will work on it and add tests.

File: test/NugetFeedDeploymentRepositoryTest/NugetFeedApplicationRepositoryTest.cs (hunk 18 above)
Comment (nehmebilal) [reply test/NugetFeedDeploymentRepositoryTest/NugetFeedApplicationRepositoryTest.cs]: I don't insist :) sounds reasonable.

File: src/NugetFeedDeploymentRepository/NugetFeedApplicationRepository.cs [hunk 23]
```diff
@@ -0,0 +1,93 @@
...
+            PackageDownloadContext downloadContext = new PackageDownloadContext(new SourceCacheContext(), tempPath, directDownload: true);
+
+            DownloadResourceResult result = await downloadResource.GetDownloadResourceResultAsync(packageIdentity, downloadContext, tempPath, _logger, CancellationToken.None);
+
+            if (result.Status != DownloadResourceResultStatus.Available)
+            {
+                throw new BinariesNotFoundException($"NuGet package for application {appIdentity} is not available from feed {this.FeedUrl}");
+            }
+
+            NugetPackageExtractor extractor = new NugetPackageExtractor();
```
Comment (nehmebilal) [inline src/NugetFeedDeploymentRepository/NugetFeedApplicationRepository.cs]: Could be a field as well or even injected to facilitate testing.

File: src/NugetFeedDeploymentRepository/NugetPackageExtractor.cs (hunk 12 above)
Comment (nehmebilal) [reply src/NugetFeedDeploymentRepository/NugetPackageExtractor.cs]: Ok no problem. I'd consider putting it behind an interface in that case (it's more than a simple POCO).

File: test/NugetFeedDeploymentRepositoryTest/NugetFeedApplicationRepositoryTest.cs [hunk 24]
```diff
@@ -0,0 +1,68 @@
...
+using Etg.Yams.Storage;
+using Etg.Yams.Utils;
+using Xunit;
+
+namespace Etg.Yams.Nuget.Storage.Test
+{
+    public class NugetFeedDeploymentRepositoryTest
+    {
+        [Fact]
+        public async Task TestHasBinaries()
```
Comment (nehmebilal) [inline test/NugetFeedDeploymentRepositoryTest/NugetFeedApplicationRepositoryTest.cs]: Add tests for ConflictResolutionMode? :)

File: Etg.Yams.sln (hunk 22 above)
Comment (jkonecki) [reply Etg.Yams.sln]: Created a nuget package and fluent YamsServiceHost builder.

Comment (nehmebilal) [reply Etg.Yams.sln]: @jkonecki I am very confused about the pattern you are using for fluent builder, which I haven't seen before. Usually, you create a builder class (e.g. YamsHostBuilder) that takes all required parameters in the constructor. Then you add fluent interface for optional configuration (take a look at YamsConfigBuilder). For external modules, they just add extension method to YamsHostBuilder.

**Comment by jkonecki** [comment]
I can do it like that but it means you don't have a fully fluent interface
(as you're using constructor).
I use the same pattern in https://github.com/aliostad/RandomGen/

**Comment by nehmebilal** [comment]
Yes please, I prefer the builder design pattern style and it's familiar and consistent for most developers. I like the constructor for required parameters (which is part of the pattern) because it forces the user to pass in everything required (as opposed to getting a runtime exception telling him that a parameter is missing).

File: Etg.Yams.sln (hunk 22 above)
Comment (jkonecki) [reply Etg.Yams.sln]: I removed the fluent interface implementation and created an overload for static `Create` method that accepts individual interfaces. 
I feel a contradiction in your requirements: on one hand you say that the constructor should take all required parameters and the extension methods should be for optional ones, and on the other hand you suggest creating an extension method for registering NuGet repository. But deployment repository is not an optional parameter (I cannot construct Yams service without one).

Comment (nehmebilal) [reply Etg.Yams.sln]: @jkonecki You're right, sorry about that. I didn't notice that all parameters were required when I suggested a builder as an option. What you have right now looks pretty good. 

There is a common variation of the builder pattern where no parameters are given in the constructor and the Build() method does the work of making sure that all required parameters have been given. This is especially useful in languages where named parameters are not supported (for readability) or when one need to construct the builder in multiple steps (pass it around). I believe ASP.Net opted for this variation. I'd happy with this variation if you prefer.

Sorry again for the confusion!

File: src/NugetFeedDeploymentRepository/NugetFeedDeploymentRepository.nuspec [hunk 25]
```diff
@@ -0,0 +1,15 @@
+<?xml version="1.0"?>
+<package >
+  <metadata>
+    <id>Etg.Yams.Client.Nuget</id>
+    <version>1.0.0</version>
+    <title>YAMS Client - Nuget extensions</title>
```
Comment (nehmebilal) [inline src/NugetFeedDeploymentRepository/NugetFeedDeploymentRepository.nuspec]: YAMS Client?

File: src/NugetFeedDeploymentRepository/NugetFeedDeploymentRepository.nuspec [hunk 26]
```diff
@@ -0,0 +1,15 @@
...
+    <authors>Microsoft</authors>
+    <owners>Microsoft</owners>
+    <licenseUrl>https://opensource.org/licenses/MIT</licenseUrl>
+	<projectUrl>https://github.com/Microsoft/Yams</projectUrl>
+    <requireLicenseAcceptance>false</requireLicenseAcceptance>
+    <description>YAMS Client can be used by YAMS application to take advantage of health monitoring and graceful shutdown features.</description>
```
Comment (nehmebilal) [inline src/NugetFeedDeploymentRepository/NugetFeedDeploymentRepository.nuspec]: Wrong description?

**Comment by nehmebilal** [comment]
@jkonecki Is this PR still relevant?
